File: languages/p/python/lambda-core.py
# LOGIC
_true = lambda x: lambda y: x
_false = lambda x: lambda y: y
_not = lambda b: b(_false)(_true)
_and = lambda b1: lambda b2: b1(b2)(_false)
_or = lambda b1: lambda b2: b1(_true)(b2)

# CHURCH NUMERALS
_zero = lambda f: lambda x: x
_succ = lambda n: lambda f: lambda x: f(n(f)(x))
_pred = lambda n: lambda f: lambda x: n(lambda g: lambda h: h(g(f)))(lambda u: x)(lambda a: a)
_one = _succ(_zero)

# HELPERS - not pure lambda calculus
read_bool = lambda b: print(b("t")("f"))
read_church = lambda n: print(n(lambda x: x+1)(0))

# EXAMPLES
print("LOGIC")
print("-------------")
print("TRUE/FALSE")
read_bool(_true)
read_bool(_false)

print("NOT")
read_bool(_not(_true))
read_bool(_not(_false))

print("AND")
read_bool(_and(_true)(_true))
read_bool(_and(_true)(_false))
read_bool(_and(_false)(_true))
read_bool(_and(_false)(_false))

print("OR")
read_bool(_or(_true)(_true))
read_bool(_or(_true)(_false))
read_bool(_or(_false)(_true))
read_bool(_or(_false)(_false))

print()
print("CHURCH NUMERALS")
print("-------------")
print("ZERO/SUCC")
read_church(_zero)
read_church(_one)
read_church(_succ(_one))
read_church(_succ(_succ(_one)))

print("PRED")
read_church(_pred(_one))
read_church(_pred(_succ(_one)))
read_church(_pred(_succ(_succ(_one))))

File: languages/p/perl/lambda-core.pl
use strict;
use warnings;

# LOGIC
my $_true  = sub { 
    # shift is more convenient
    # my $x = $_[0];
    my $x = shift; 
    sub { $x } 
};
my $_false = sub { 
    sub { shift } 
};

my $not = sub { 
    my $b = shift; 
    $b->($_false)->($_true) 
};

my $and = sub {
    my $b1 = shift;
    sub {
        my $b2 = shift;
        $b1->($b2)->($_false)
    }
};

my $or = sub {
    my $b1 = shift;
    sub {
        my $b2 = shift;
        $b1->($_true)->($b2)
    }
};

# CHURCH NUMERALS
my $zero = sub { sub { shift } };
my $succ = sub {
    my $n = shift;
    sub {
        my $f = shift;
        sub {
            my $x = shift;
            $f->($n->($f)->($x))
        }
    }
};
my $pred = sub {
    my $n = shift;
    sub {
        my $f = shift;
        sub {
            my $x = shift;
            $n->(sub {
                my $g = shift;
                sub {
                    my $h = shift;
                    $h->($g->($f))
                }
            })->(sub { $x })->(sub { shift })
        }
    }
};
my $one = $succ->($zero);

# HELPERS - not pure lambda calculus
sub read_bool {
    my $b = shift;
    print $b->('true')->('false');
}

sub read_church {
    my $n = shift;
    print $n->(sub { $_[0] + 1 })->(0);
}

# EXAMPLES
print "LOGIC\n";
print "-------------\n";
print " TRUE: ";
read_bool($_true);
print "\n FALSE: ";
read_bool($_false);

print "\n NOT TRUE: ";
read_bool($not->($_true));
print "\n NOT FALSE: ";
read_bool($not->($_false));

print "\n TRUE AND TRUE: ";
read_bool($and->($_true)->($_true));
print "\n TRUE AND FALSE: ";
read_bool($and->($_true)->($_false));
print "\n FALSE AND TRUE: ";
read_bool($and->($_false)->($_true));
print "\n FALSE AND FALSE: ";
read_bool($and->($_false)->($_false));

print "\n TRUE OR TRUE: ";
read_bool($or->($_true)->($_true));
print "\n TRUE OR FALSE: ";
read_bool($or->($_true)->($_false));
print "\n FALSE OR TRUE: ";
read_bool($or->($_false)->($_true));
print "\n FALSE OR FALSE: ";
read_bool($or->($_false)->($_false));

print "\n\nCHURCH NUMERALS\n";
print "-------------\n";
print " ZERO: ";
read_church($zero);
print "\n ONE: ";
read_church($one);
print "\n SUCC ONE: ";
read_church($succ->($one));
print "\n SUCC SUCC ONE: ";
read_church($succ->($succ->($one)));

print "\n PRED ONE: ";
read_church($pred->($one));
print "\n PRED SUCC ONE: ";
read_church($pred->($succ->($one)));
print "\n PRED SUCC SUCC ONE: ";
read_church($pred->($succ->($succ->($one))));
